### PR TITLE
[Core] Handle space removed on no rules applied on PhpFileProcessor 

### LINKED ISSUE #7748: Can we revert file changes if no rules applied?
# Question

I know that, due to AST, there is a small subset of files where lines are trimmed. 

This is not a problem if project is small and not if there are other changes in the file.

But I'm facing a huge >20k file project. Running a simple custom rule to do a name change to fix a typo,  I got over 2400 files changed, but only 24 with the fix applied.

So, I am asking if there is something like a 'postFileProcessing' hook where to count file changes, or verify if any rules applied, and then ask rector to simply NOT save the file. 

This avoid the problem of "useless" trimming of files. 

I also known is an already spoken question, but I'm posing it in a different way, to eventually ipotize a different solution.

Also, an hook fired after each file processed could be very usefull for A LOT of automations

### PR DESCRIPTION
**Pull request by samsonasik**
@staabm @Mirko-Tebaldi @realtebo here another try to ensure no space removal on no rules applied.

Fixes https://github.com/rectorphp/rector/issues/7748

### CODE REVIEW

**Comment by samsonasik** [comment]
It seems cause test error on `DowngradeAnonymousClassRectorTest`

```diff
There were 4 failures:

1) Rector\Tests\DowngradePhp70\Rector\New_\DowngradeAnonymousClassRector\DowngradeAnonymousClassRectorTest::test with data set #3
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
 <?php
 
+
 function inFunctionNoNamespace()
 {
-    return new Anonymous__%s__0();
+    return new Anonymous__ca71b91d0b8bac03aee8f01071b87f9a__0();
 }
-class Anonymous__%s__0
+class Anonymous__ca71b91d0b8bac03aee8f01071b87f9a__0
```

Ref https://github.com/rectorphp/rector-src/actions/runs/4269346270/jobs/7432468700#step:7:31

**Comment by samsonasik** [comment]
Should be fixed now 👍

**Comment by samsonasik** [comment]
Finally 🎉 🎉 🎉 All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by samsonasik** [comment]
Also works on multiple file extensions 🎉 

```php
$rectorConfig->fileExtensions(['php', 'phtml']);
```

https://github.com/rectorphp/rector-src/pull/3412/commits/590415f7a6f0f828e088ddf8c7c94f6a8a58b714 
https://github.com/rectorphp/rector-src/pull/3412/commits/ded9a54b9bef8820c353f88a11d589379e1458b2

File: src/PhpParser/Printer/FormatPerservingPrinter.php [hunk 1]
```diff
@@ -56,4 +53,12 @@ public function printParsedStmstAndTokens(File $file): string
             $file->getOldTokens()
         );
     }
+
+    public function dumpFile(string $filePath, string $newContent): void
+    {
+        $this->filesystem->dumpFile($filePath, $newContent);
+
+        // @todo how to keep origianl access rights without the SplFileInfo
```
Comment (staabm) [inline src/PhpParser/Printer/FormatPerservingPrinter.php]: debug leftover?

Comment (samsonasik) [reply src/PhpParser/Printer/FormatPerservingPrinter.php]: it was the original code, I think @TomasVotruba was trying to make not rely on `Filesystem` with different implementation and mark as todo :). 

I just moved the functionality to new method as used to re-use on rollback original file content.

**Comment by samsonasik** [comment]
@TomasVotruba I removed `FormatPerservingPrinter::printParsedStmstAndTokens()` to avoid unnecessary rollback file content, just return early after compare the string. The functionality replaced with next 

```php
        if (! $configuration->isDryRun()) {
            $this->formatPerservingPrinter->dumpFile($file->getFilePath(), $newContent);
        }
```

with `$newContent` is the string printed.

File: src/PhpParser/Printer/FormatPerservingPrinter.php (hunk 1 above)
Comment (TomasVotruba) [reply src/PhpParser/Printer/FormatPerservingPrinter.php]: I think there was also same issue that file was reprinted always, but with different than orignal rights.
Now it's gone I assume as nobody filled an issue in this absence :)

Comment (samsonasik) [reply src/PhpParser/Printer/FormatPerservingPrinter.php]: @TomasVotruba do you think commented code can be removed?

Comment (samsonasik) [reply src/PhpParser/Printer/FormatPerservingPrinter.php]: @staabm @TomasVotruba I removed commented code https://github.com/rectorphp/rector-src/pull/3412/commits/c3a76176710d97e517f7d85c61a9566b4f0095a4 👍

**Comment by TomasVotruba** [comment]
@samsonasik Keep mergin these when ready :) I think it's your know-how domain now :+1: :clap:

**Pull request merged**